types/basetypes/dynamic_value.go

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package basetypes
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+var (
+	_ DynamicValuable = DynamicValue{}
+)
+
+// DynamicValuable extends attr.Value for dynamic value types. Implement this interface
+// to create a custom Dynamic value type.
+type DynamicValuable interface {
+	attr.Value
+
+	// ToDynamicValue should convert the value type to a DynamicValue.
+	ToDynamicValue(context.Context) (DynamicValue, diag.Diagnostics)
+}
+
+// DynamicValuableWithSemanticEquals extends DynamicValuable with semantic equality logic.
+type DynamicValuableWithSemanticEquals interface {
+	DynamicValuable
+
+	// DynamicSemanticEquals should return true if the given value is
+	// semantically equal to the current value. This logic is used to prevent
+	// Terraform data consistency errors and resource drift where a value change
+	// may have inconsequential differences.
+	//
+	// Only known values are compared with this method as changing a value's
+	// state implicitly represents a different value.
+	DynamicSemanticEquals(context.Context, DynamicValuable) (bool, diag.Diagnostics)
+}
+
+// NewDynamicValue creates a Dynamic with a known value. Access the value via the Dynamic
+// type UnderlyingValue method. The concrete value type returned to Terraform from this value
+// will be determined by the provided `(attr.Value).ToTerraformValue` function.
+func NewDynamicValue(value attr.Value) DynamicValue {
+	return DynamicValue{
+		value: value,
+		state: attr.ValueStateKnown,
+	}
+}
```

Comment:
Another place for feedback, should we validate the `value` being passed in here? To ensure no one tries to do something like:

```go
// Will cause errors eventually
types.DynamicValue(nil)

// Will cause errors eventually
types.DynamicValue(types.Dynamic{})
```

We could follow the collection/object pattern of `types.DynamicValue` returning a diagnostic and then having a `types.DynamicValueMust`

Reply:
I think we should go ahead and do what you propose here! Better to catch/report the problem(s) upfront rather than later.

Reply:
Here is the proposal with some tests: 81afd88, lmk what you think

Reply:
Well the fun part (depending on what you think is fun 🙃)  about how it's implemented right now.... because it's all a pass-through container it'll just work anyways lol

```go
// It's double nested :O
data.DynamicComputed = types.DynamicValue(
    types.DynamicValue(
        types.ListValueMust(
            types.Int64Type,
            []attr.Value{
                types.Int64Value(1),
                types.Int64Value(2),
                types.Int64Value(3),
                types.Int64Value(4),
                types.Int64Value(5),
            },
        ),
    ),
)
```

```json
{
  "dynamic_computed": {
    "value": [
      1,
      2,
      3,
      4,
      5
    ],
    "type": [
      "list",
      "number"
    ]
  },
}
```

Reply:
That _is_ fun! I'll resolve this then.